service/history/workflow/mutable_state_impl.go

```diff
@@ -6927,21 +6927,46 @@ func (ms *MutableStateImpl) applyUpdatesToStateMachineNodes(
 				// we don't have enough information to recreate all parents
 				return err
 			}
-			newNode, err := parentNode.AddChild(incomingPath[len(incomingPath)-1], nodeMutation.Data)
-			if err != nil {
-				return err
+
+			key := incomingPath[len(incomingPath)-1]
+			parentInternalNode := parentNode.InternalRepr()
+			machines, ok := parentInternalNode.Children[key.Type]
+			if ok {
+				if _, ok = machines.MachinesById[key.ID]; ok {
+					if ok {
+						return fmt.Errorf("%w: %v", hsm.ErrStateMachineAlreadyExists, key)
```

Comment:
This should never happen, since you know that you're handling `ErrStateMachineNotFound` in this branch. You can just continue to where you add the node to the map below.